---
title: "Infer ASVs with DADA2"
author: "Gurpreet Kaur"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center",
                      fig.path = "../figures/01_DADA2/") #send any figure output to this folder
```

# Goals of this file

1. Use raw fastq files and generate the quality plots to assess quality of reads.
2. Filter and trim out bad sequences and bases from our sequencing files.
3. Write out fastq files with high quality sequences.
4. Evaluate the quality from our filter and trim.

#Load libraries
```{r load-libraries}
#install.packages("devtools")
library(devtools)

#devtools::install_github("benjjneb/dada2")
library(dada2)

```




#Load Data
```{r load-data}
#set the raw fastq path to the raw sequencing files
#path to fastq files

raw_fastqs_path <- "data/01_DADA2/01_raw_gzipped_fastqs"
raw_fastqs_path

#what files are in this path? Intuition check!
list.files(raw_fastqs_path)

#how many files?
str(list.files(raw_fastqs_path))

#create a vector of forward reads
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE)

head(forward_reads)


#create vector of reverse reads
reverse_reads <- list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)
head(reverse_reads)

```

#Quality plots
```{r raw-quality-plot}
#randomly select two samples from dataset to evaluate
random_samples <- sample(1:length(reverse_reads), size = 2)
random_samples

#calculate and plot quality of these two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])
```

#Prepare aplace holder for filtered reads
```{r prep-filtered-sequences}
#Place filtered reads into filtered_fastqs_path
samples <- sapply(strsplit(basename(forward_reads), "_"), `[`,1)
head(samples)

filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"

#create two variables: filtered_F, filtered_R
filtered_forward_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R1_filtered.fastq.gz"))
length(filtered_forward_reads)


filtered_reverse_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R2_filtered.fastq.gz"))
length(filtered_reverse_reads)

head(filtered_reverse_reads)


```




#Filter and trim reads
```{r filter-and-trim}
#?filterAndTrim



```

