---
title: "Phylogenetic Tree Construction"
author: "Gurpreet Kaur"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.path = "../figures/04_Biodiversity")
```


#Goals
To create a phylogenetic tree!
1. Load in preprocessed phyloseq object
2. Create ASV fasta file from the phyloseq object
3. Align the 16S sequences from the fasta file with MAFFT
4. Create a tree with FastTree2


##Set the seed
```{r}
set.seed(238428)
```


##Load packages and phyloseq object
```{r load-environment}
#phytools, ggtree, RColorBrewer
pacman::p_load(phytools, ggtree, RColorBrewer, install = FALSE)

#Load phyloseq object
load("data/02_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq
```

#Create fasta file of ASV and their sequences

This fasta file will be used to create our alignment in MAFFT
```{r}
#pull out ASV seqs and ASV names

asv_seq_df <-
  raw_preprocessed_physeq@tax_table %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq)

# Add the > to make fasta header
asv_seq_df$ASV <- paste0(">", asv_seq_df$ASV)

#create the fasta object
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASVseq))
head(asv_seq_fasta)

#Write to a file
write(asv_seq_fasta, file = "data/03_Phylogentic_Tree/preprocessed_ASVs.fasta")
```


#Align the 16S sequences from fasta file with MAFFT

```{r run-mafft, engine = 'bash', engine.opts = '-l'}
#write bash code to run mafft
#first provide path to mafft
export PATH=/programs/mafft/bin:$PATH

#change working directories to provide the fasta file we made above
cd data/03_Phylogentic_Tree/
pwd 

#set a seed - using same seed for consistency
RANDOM=238428

#Run mafft
#for now, use default options, not the version for reproducibility
#mafft automatically knows that it is a nucleotide alignment

/programs/mafft/bin/mafft --auto preprocessed_ASVs.fasta > MAFFT_aligned_ASVs.fasta

#Change back to the project directory
cd ../../
pwd

```


#Session information
```{r}
#ensure reproducibility
devtools::session_info()
```

